src/coord/src/coord.rs

```diff
@@ -2664,13 +2675,45 @@ impl Coordinator {
         //     since.join_assign(&self.source_info[instance_id].since);
         // }
 
+        for (source_id, _) in dataflow.source_imports.iter() {
+            let source_timeline = self
+                .timelines
+                .get(source_id)
+                .expect("global id missing at coord");
+
+            if timeline.is_none() {
+                timeline = Some(*source_timeline);
+            }
+
+            let timeline = timeline.expect("known to exist");
+
+            if timeline != *source_timeline {
+                coord_bail!("Dataflow requested sources spanning multiple timelines");
+            }
+        }
+
         // For each imported arrangement, lower bound `since` by its own frontier.
         for (global_id, (_description, _typ)) in dataflow.index_imports.iter() {
             since.join_assign(
                 self.indexes
                     .since_of(global_id)
                     .expect("global id missing at coordinator"),
             );
+
+            let index_timeline = self
```

Comment:
Can this be de-duplicated? It's unfortunate we need to check for this manually (since we could miss places where it needs to be checked), but this is probably easier than changing the timestamp type for now.